frame/vamm/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+	pub struct GenesisConfig<T: Config> {
+		pub vamm_count: VammIdOf<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { vamm_count: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			VammCounter::<T>::put(self.vamm_count);
+		}
+	}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                           Vamm Trait
+	// ----------------------------------------------------------------------------------------------------
+
+	impl<T: Config> Vamm for Pallet<T> {
+		type VammId = VammIdOf<T>;
+		type Balance = BalanceOf<T>;
+
+		/// Creates a new virtual market.
+		///
+		/// # Overview
+		/// In order for the caller to create new markets, it has to request it
+		/// to the VAMM, which is responsible to keep track of all active
+		/// markets. The VAMM creates the market, inserts it in storage,
+		/// deposits a [`Created`](Event::<T>::Created) event on the
+		/// blockchain and returns the new market ID to the caller.
+		///
+		/// In the diagram below the Clearing House is depicted as the caller.
+		///
+		/// ![](https://www.plantuml.com/plantuml/svg/TP3FQi904CRl-nHZAbIHWCMRs81_3kt1Ndo2SJ8cYybEkZk9hzziIYeMtDF2z_j-0r-uMjUWnneyXqPSu2E7W0Nlk9BRrdkvWVgMZLgj6BhjyGWr-Yiha6TKAmvIEBL4VGqkVSUyOgk2fBP3PH1d9bfopR4MpAJnGfotdc5wGLlrdzcq3iNS06mkudgjLEBVanUYPV-IR7FE8c0cxFA_yeCd_5v_ubckjlktrJEFQT2h9TjWNqds5QEthe0FQGCdW06eV4JY0aFGOLsR71NF61YIauh7Wc4MWVb0X7_8hXAwKedM3V6PZA4IqcnGBHPhsCT5UTpNytVBGKrC8pNe7g4nJ3FfTMiuS2F1Aj30vAE9EtPt5AXCq_LzjkIBRoFvUKZcvWS0)
+		///
+		/// ## Parameters:
+		/// - `base_asset_reserves`: The amount of base asset
+		/// - `quote_asset_reserves`: The amount of quote asset
+		/// - `peg_multiplier`: The constant multiplier responsible to balance quote and base asset
+		///
+		/// ## Returns
+		/// The new virtual market id, if successful.
+		///
+		/// ## Assumptions or Requirements
+		/// TODO
+		///
+		/// ## Emits
+		/// * [`Created`](Event::<T>::Created)
+		///
+		/// ## State Changes
+		/// Updates [`VammMap`] storage map and [`VammCounter`] storage value.
+		///
+		/// ## Errors
+		/// * [`ArithmeticError::Overflow`](sp_runtime::ArithmeticError)
+		///
+		/// # Runtime
+		/// `O(1)`
+
+		#[transactional]
+		fn create(
+			base_asset_reserves: Self::Balance,
+			quote_asset_reserves: Self::Balance,
+			peg_multiplier: Self::Balance,
+		) -> Result<Self::VammId, DispatchError> {
+			// TODO: (Matheus)
+			// How to ensure that the caller has the right privileges?
+			// (eg. How to ensure the caller is the Clearing House, and not anyone else?)
+			ensure!(!base_asset_reserves.is_zero(), Error::<T>::BaseAssetReserveIsZero);
+			ensure!(!quote_asset_reserves.is_zero(), Error::<T>::QuoteAssetReserveIsZero);
+			ensure!(!peg_multiplier.is_zero(), Error::<T>::PegMultiplierIsZero);
```

Comment:
Good call verifying the inputs before the `try_mutate` call; looks cleaner this way. I think I'll do that too

Reply:
You should try to have all validity checks occur before any state mutations

Reply:
Yeah, but it seems that the `try_mutate` function will not mutate the corresponding storage item if any error is thrown. At least, in my implementation I checked through tests that the storage is left intact